### PR TITLE
HBASE-25420 Some minor improvements in rpc implementation

### PR DESCRIPTION
**Pull request by Apache9**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 21s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 53s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 35s |  master passed  |
| +1 :green_heart: |  spotbugs  |   3m  6s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 33s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 29s |  hbase-client: The patch generated 0 new + 0 unchanged - 3 fixed = 0 total (was 3)  |
| +1 :green_heart: |  checkstyle  |   1m  7s |  hbase-server: The patch generated 0 new + 10 unchanged - 10 fixed = 10 total (was 20)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  18m  5s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1 3.3.0.  |
| +1 :green_heart: |  spotbugs  |   3m 27s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 25s |  The patch does not generate ASF License warnings.  |
|  |   |  44m 42s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2792 |
| Optional Tests | dupname asflicense spotbugs hadoopcheck hbaseanti checkstyle |
| uname | Linux 7e870a687e6f 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f600856a3b |
| Max. process+thread count | 94 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=3.1.12 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by ddupg — APPROVED** [review]
+1

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 29s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 21s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m 10s |  master passed  |
| +1 :green_heart: |  compile  |   1m 34s |  master passed  |
| +1 :green_heart: |  shadedjars  |   6m 45s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  8s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 34s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 34s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   6m 41s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  6s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m  9s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 132m 18s |  hbase-server in the patch passed.  |
|  |   | 163m 54s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2792 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux b1ffc912ffea 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f600856a3b |
| Default Java | AdoptOpenJDK-11.0.6+10 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/testReport/ |
| Max. process+thread count | 4601 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 28s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 37s |  master passed  |
| +1 :green_heart: |  compile  |   1m 24s |  master passed  |
| +1 :green_heart: |  shadedjars  |   6m 36s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  1s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 30s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 21s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 21s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   6m 37s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 58s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m  7s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 139m  6s |  hbase-server in the patch passed.  |
|  |   | 168m 51s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2792 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux d79f5eb44b1f 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f600856a3b |
| Default Java | AdoptOpenJDK-1.8.0_232-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/testReport/ |
| Max. process+thread count | 4019 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2792/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java [hunk 1]
```diff
@@ -103,8 +103,8 @@ private void writeRequest(ChannelHandlerContext ctx, Call call, ChannelPromise p
         ctx.write(buf, withoutCellBlockPromise);
         ChannelPromise cellBlockPromise = ctx.newPromise();
         ctx.write(cellBlock, cellBlockPromise);
-        PromiseCombiner combiner = new PromiseCombiner();
-        combiner.addAll(withoutCellBlockPromise, cellBlockPromise);
+        PromiseCombiner combiner = new PromiseCombiner(ctx.executor());
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java:106]: This a bug fix? Passing the executor?

File: hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java [hunk 2]
```diff
@@ -103,8 +103,8 @@ private void writeRequest(ChannelHandlerContext ctx, Call call, ChannelPromise p
...
+        combiner.addAll((ChannelFuture) withoutCellBlockPromise, cellBlockPromise);
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java:107]: We need a cast is needed so we call a subclass method ?

**Review by saintstack — APPROVED** [review]
LGTM. Just some bit I don't get.. so a question.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java (hunk 1 above)
Comment (Apache9) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java:106]: The default constructor has been deprecated.

https://github.com/netty/netty/blob/3e8e52725bbd9d999873880956bc8f241372daa8/common/src/main/java/io/netty/util/concurrent/PromiseCombiner.java#L73

File: hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java (hunk 2 above)
Comment (Apache9) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcDuplexHandler.java:107]: The method which takes Promises as parameters is deprecated.

https://github.com/netty/netty/blob/3e8e52725bbd9d999873880956bc8f241372daa8/common/src/main/java/io/netty/util/concurrent/PromiseCombiner.java#L123

So here we cast the first one to a future to let it call the method which takes Futures as parameters(Promise is a sub interface of Future).

**Pull request merged**